### PR TITLE
Fix AsVersionedObject to deal with external schema

### PR DESCRIPTION
**Pull request by mfojtik**
This fixes errors where the internal object omits json tags or has a different field structure than the versioned object

At also add typer and encoder to `AsVersionedObject()` to deal with custom schemes. 

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36066)

### CODE REVIEW

**Comment by mfojtik** [comment]
@kargakis @smarterclayton PTAL

@soltysh 🍺

**Comment by soltysh** [comment]
LGTM

**Comment by kargakis** [comment]
@kubernetes/sig-api-machinery @kubernetes/sig-cli

**Comment by mfojtik** [comment]
(need to add unit test)

**Comment by mfojtik** [comment]
(added unit test)

**Comment by soltysh** [comment]
@liggitt || @deads2k any objections for merge at this point?

**Comment by liggitt** [comment]
talked with @mfojtik, would like to see a test that fails without the fix in place first

**Comment by mfojtik** [comment]
@liggitt now that I spend couple hours making that test working, I think I would defer this as making this work in kube will require more work.

Basically, when I register new schema and add my types there, the `AsVersionedObject` method have `api.Schema.ObjectKinds(info.Object)` hardcoded, so my custom types will get converted to `runtime.Unknown` and the test will fail....
This also explains why this fix is working just fine for origin (Origin types are registered in `api.Schema`?).

To make the test work (I can show you WIP) we will need to add `runtime.ObjectTyper` to `AsVersionedObject` (or even runtime.Schema) and plumb this all the way up to factory and all tests...

So what I suggest is to merge this as it is now and open a follow up to refactor this post-mortem. We also need this for Origin to fix the regression bug which is due to tomorrow. WDYT?

**Comment by mfojtik** [comment]
@liggitt the WIP of the test is here: 

https://github.com/mfojtik/kubernetes/commit/2c351afd51eff4f8bfe1ee8ad6394f93b96a7479

I can submit this with the test disabled until we fix the `AsVersionedObject (https://github.com/kubernetes/kubernetes/blob/56bbfd259a6f45dfbcfb6f3c9f61643a787ed067/pkg/kubectl/resource/result.go#L254)

**Comment by smarterclayton** [comment]
The problem here is the use of Get and the structured scheme. Get rid of /
fix either and you fix this issue.

On Nov 3, 2016, at 9:59 AM, Michal Fojtik <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> the WIP of the test is here:

mfojtik@2c351af
<https://github.com/mfojtik/kubernetes/commit/2c351afd51eff4f8bfe1ee8ad6394f93b96a7479>

I can submit this with the test disabled until we fix the
`AsVersionedObject (
https://github.com/kubernetes/kubernetes/blob/56bbfd259a6f45dfbcfb6f3c9f61643a787ed067/pkg/kubectl/resource/result.go#L254
)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/36066#issuecomment-258149604>,
or mute the thread

**Comment by mfojtik** [comment]
@smarterclayton i'm not sure I understand what you're suggesting

**Comment by mfojtik** [comment]
@liggitt added typer/encoder into AsVersionedObject and added unit test.

**Comment by mfojtik** [comment]
@liggitt PTAL

File: pkg/kubectl/cmd/util/helpers_test.go [hunk 1]
```diff
@@ -329,6 +331,94 @@ func TestDumpReaderToFile(t *testing.T) {
...
+	// TestType is an internal type without JSON tags
+	internalGV := unversioned.GroupVersion{Group: "test", Version: runtime.APIVersionInternal}
+	externalGV := unversioned.GroupVersion{Group: "test", Version: "v1"}
+
+	s := runtime.NewScheme()
+	s.AddKnownTypes(internalGV, &TestType{})
+	s.AddKnownTypes(externalGV, &ExternalTestType{})
+	s.AddKnownTypeWithName(externalGV.WithKind("TestType"), &ExternalTestType{})
+	s.AddKnownTypeWithName(internalGV.WithKind("TestType"), &TestType{})
+	// TODO: Use AddToScheme(s) here instead of explicitely calling the conversion function.
```
Comment (mfojtik) [inline pkg/kubectl/cmd/util/helpers_test.go]: @liggitt if I do this I will get:

```
unexpected error: converting (util.TestType).CreationTimestamp.Time.sec to (util.ExternalTestType).CreationTimestamp.Time.sec: Cannot set dest. (Tried to deep copy something with unexported fields?)
```

**Comment by soltysh** [comment]
@k8s-bot node e2e test this

**Comment by soltysh** [comment]
Still LGTM

File: pkg/kubectl/cmd/edit.go [hunk 2]
```diff
@@ -340,10 +340,10 @@ func getPrinter(cmd *cobra.Command) (*editPrinterOptions, error) {
 	}
 }
 
-func getMapperAndResult(f cmdutil.Factory, args []string, options *resource.FilenameOptions, editMode EditMode) (meta.RESTMapper, *resource.Mapper, *resource.Result, string, error) {
+func getMapperAndResult(f cmdutil.Factory, args []string, options *resource.FilenameOptions, editMode EditMode) (meta.RESTMapper, *resource.Mapper, runtime.ObjectTyper, *resource.Result, string, error) {
```
Comment (liggitt) [inline pkg/kubectl/cmd/edit.go]: don't you already return the typer in `resourceMapper.ObjectTyper`?

Comment (mfojtik) [inline pkg/kubectl/cmd/edit.go]: you're right! will fix it

File: pkg/kubectl/cmd/get_test.go [hunk 3]
```diff
@@ -225,10 +225,10 @@ func TestGetUnknownSchemaObjectListGeneric(t *testing.T) {
 		}
 		arr := out["items"].([]interface{})
 		if arr[0].(map[string]interface{})["apiVersion"] != test.testtypeVersion {
```
Comment (liggitt) [inline pkg/kubectl/cmd/get_test.go]: save locally and reuse, rather than repeating the index+cast:
```
if version := arr[0].(map[string]interface{})["apiVersion"]; version != test.testtypeVersion {
  ... version ...
}
```

Comment (mfojtik) [inline pkg/kubectl/cmd/get_test.go]: i can drop this change, just something I encountered and was hard to debug.

File: pkg/kubectl/cmd/get.go [hunk 4]
```diff
@@ -150,6 +150,9 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 	allNamespaces := cmdutil.GetFlagBool(cmd, "all-namespaces")
 	showKind := cmdutil.GetFlagBool(cmd, "show-kind")
 	mapper, typer := f.Object()
```
Comment (mfojtik) [inline pkg/kubectl/cmd/get.go]: ?

(I need to expose typer in GetOptions for a unit test)

Comment (liggitt) [inline pkg/kubectl/cmd/get.go]: it's possible the test typer we mocked up in NewMixedFactory is wrong, or it's possible something else is wrong, but we need to figure out the problem there and fix it, not swap in a fake typer here for test

File: pkg/kubectl/cmd/get_test.go [hunk 5]
```diff
@@ -210,7 +210,7 @@ func TestGetUnknownSchemaObjectListGeneric(t *testing.T) {
 		cmd.Flags().Set("output", "json")
 
 		cmd.Flags().Set("output-version", test.outputVersion)
-		err := RunGet(f, buf, errBuf, cmd, []string{"type/foo", "replicationcontrollers/foo"}, &GetOptions{})
+		err := RunGet(f, buf, errBuf, cmd, []string{"type/foo", "replicationcontrollers/foo"}, &GetOptions{Typer: api.Scheme})
```
Comment (liggitt) [inline pkg/kubectl/cmd/get_test.go]: we need to revert this change and figure out whether the problem is in the typer returned from NewMixedFactory, or in AsVersionedObject, or elsewhere

File: pkg/kubectl/cmd/get.go (hunk 4 above)
Comment (liggitt) [inline pkg/kubectl/cmd/get.go]: the test MixedFactory was actually returning a typer that would recognize the test types, but the point of `TestGetUnknownSchemaObjectListGeneric` is to test serialization of objects the compiled-in typer *doesn't* recognize. Because you plumbed the typer correctly, we needed a way to simulate the typer not recognizing the test type. 

fixed in https://github.com/mfojtik/kubernetes/pull/1, and removed this field and test-only hack

**Comment by mfojtik** [comment]
@liggitt I think @deads2k already took care of the "get" here: 

https://github.com/kubernetes/kubernetes/pull/36085 

So no plumbing is needed with this PR.

**Comment by soltysh** [comment]
It's possible that this is fixed in #36369 much more cleanly. @mfojtik mind checking that?

**Comment by deads2k** [comment]
> It's possible that this is fixed in #36369 much more cleanly. @mfojtik mind checking that?

That ended up too big for 1.5

**Comment by liggitt** [comment]
yeah, #36085 fixed the get case

**Comment by mfojtik** [comment]
@liggitt @deads2k you OK merging this as it is?
